src/ExpressionEvaluator/CSharp/Source/ExpressionCompiler/Symbols/EEMethodSymbol.cs

```diff
@@ -149,11 +149,17 @@ internal sealed class EEMethodSymbol : MethodSymbol
             foreach (var pair in sourceDisplayClassVariables)
             {
                 var variable = pair.Value;
-                var displayClassInstanceFromLocal = variable.DisplayClassInstance as DisplayClassInstanceFromLocal;
-                var displayClassInstance = (displayClassInstanceFromLocal == null) ?
-                    (DisplayClassInstance)new DisplayClassInstanceFromThis(_parameters[0]) :
-                    new DisplayClassInstanceFromLocal((EELocalSymbol)localsMap[displayClassInstanceFromLocal.Local]);
-                variable = variable.SubstituteFields(displayClassInstance, this.TypeMap);
+                var oldDisplayClassInstance = variable.DisplayClassInstance;
+
+                // Note: the code path for DisplayClassInstanceFromLocal is equivalent to calling 
+                // oldDisplayClassInstance.ToOtherMethod, except that doing that would produce
+                // a new LocalSymbol that would not be ReferenceEquals to the one in this.LocalsForBinding.
```

Comment:
Useful for review, but before checking in, consider revising to something along the lines of:

Note:  Must call oldDisplayClassInstance.ToOtherMethod, so that we get a LocalSymbol that is ReferenceEquals to the one in this.LocalsForBinding.